src/pycookiecheat/common.py

```diff
@@ -57,3 +59,46 @@ def generate_host_keys(hostname: str) -> Iterator[str]:
         domain = ".".join(labels[-i:])
         yield domain
         yield "." + domain
+
+
+def _deprecation_warning(msg: str):
+    warn(msg, DeprecationWarning, stacklevel=3)
+
+
+class BrowserType(str, Enum):
+    """Provide discrete values for recognized browsers.
+
+    This utility class helps ensure that the requested browser is specified
+    precisely or fails early by matching against user input; internally this
+    enum should be preferred as compared to passing strings.
+
+    >>> "chrome" == BrowserType.CHROME
+    True
+
+    TODO: consider using `enum.StrEnum` once pycookiecheat depends on python >=
+    3.11
+    """
+
+    BRAVE = "brave"
+    CHROME = "chrome"
+    CHROMIUM = "chromium"
+    FIREFOX = "firefox"
+    SLACK = "slack"
+
+    @classmethod
+    def _missing_(cls, value: str):
+        """Provide case-insensitive matching for input values.
+
+        >>> BrowserType("chrome")
+        <BrowserType.CHROME: 'chrome'>
+        >>> BrowserType("FiReFoX")
+        <BrowserType.FIREFOX: 'firefox'>
+        >>> BrowserType("edge")
+        Traceback (most recent call last):
+        ValueError: 'edge' is not a valid BrowserType
+        """
+        folded = value.casefold()
+        for member in cls:
+            if member.value == folded:
+                return member
+        raise ValueError("%r is not a valid %s" % (value, cls.__qualname__))
```

Comment:
You're right, thanks. I was thinking of [this performance lint](https://stackoverflow.com/a/34634301) for logging (also [recommended here](https://docs.python.org/3/howto/logging.html#optimization)). Doesn't seem to work for exceptions like it does for logging.

Reply:
yeah that only works for the shape of `log('this happened: %s', what_happened)` -- because then the expansion is happening in the method, and might be skipped if e.g. the log level doesn't match. If you do `%` right away that optimization isn't possible. It has to expand at that point.

(And for exceptions it makes sense, because when you're raising it, you're already at the point where you need the expanded string -- so no needless performance penalty really.)